streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredVersionedKeyValueStore.java

```diff
@@ -0,0 +1,227 @@
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.StateStoreContext;
+import org.apache.kafka.streams.processor.internals.SerdeGetter;
+import org.apache.kafka.streams.query.Position;
+import org.apache.kafka.streams.query.PositionBound;
+import org.apache.kafka.streams.query.Query;
+import org.apache.kafka.streams.query.QueryConfig;
+import org.apache.kafka.streams.query.QueryResult;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.TimestampedKeyValueStore;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+import org.apache.kafka.streams.state.VersionedBytesStore;
+import org.apache.kafka.streams.state.VersionedKeyValueStore;
+import org.apache.kafka.streams.state.VersionedRecord;
+
+/**
+ * A metered {@link VersionedKeyValueStore} wrapper that is used for recording operation
+ * metrics, and hence its inner {@link VersionedBytesStore} implementation does not need to provide
+ * its own metrics collecting functionality. The inner {@code VersionedBytesStore} of this class
+ * is a {@link KeyValueStore} of type &lt;Bytes,byte[]&gt;, so we use {@link Serde}s
+ * to convert from &lt;K,ValueAndTimestamp&lt;V&gt&gt; to &lt;Bytes,byte[]&gt;. In particular,
+ * {@link NullableValueAndTimestampSerde} is used since putting a tombstone to a versioned key-value
+ * store requires putting a null value associated with a timestamp.
+ *
+ * @param <K> The key type
+ * @param <V> The (raw) value type
+ */
+public class MeteredVersionedKeyValueStore<K, V>
+    extends WrappedStateStore<VersionedBytesStore, K, V>
+    implements VersionedKeyValueStore<K, V> {
+
+    private final MeteredVersionedKeyValueStoreInternal internal;
+
+    MeteredVersionedKeyValueStore(final VersionedBytesStore inner,
+                                  final String metricScope,
+                                  final Time time,
+                                  final Serde<K> keySerde,
+                                  final Serde<ValueAndTimestamp<V>> valueSerde) {
+        super(inner);
+        internal = new MeteredVersionedKeyValueStoreInternal(inner, metricScope, time, keySerde, valueSerde);
+    }
+
+    /**
+     * Private helper class which represents the functionality of a {@link VersionedKeyValueStore}
+     * as a {@link TimestampedKeyValueStore} so that the bulk of the metering logic may be
+     * inherited from {@link MeteredKeyValueStore}. As a result, the implementation of
+     * {@link MeteredVersionedKeyValueStore} is a simple wrapper to translate from this
+     * {@link TimestampedKeyValueStore} representation of a versioned key-value store into the
+     * {@link VersionedKeyValueStore} interface itself.
+     */
+    private class MeteredVersionedKeyValueStoreInternal
+        extends MeteredKeyValueStore<K, ValueAndTimestamp<V>>
+        implements TimestampedKeyValueStore<K, V> {
```

Comment:
`MeteredVersionedKeyValueStoreInternal` and `MeteredTimestampKeyValueStore` use different serdes -- `MeteredTimestampKeyValueStore` uses `ValueAndTimestampSerde` while `MeteredVersionedKeyValueStoreInternal` uses `NullableValueAndTimestampSerde`. 

If you look at the code for `MeteredTimestampKeyValueStore`, the only method it overrides from `MeteredKeyValueStore` is `prepareValueSerdeForStore()`. We could have `MeteredVersionedKeyValueStoreInternal` extend `MeteredTimestampKeyValueStore` instead of `MeteredKeyValueStore` if we want but `MeteredVersionedKeyValueStoreInternal` needs to override `prepareValueSerdeForStore()` with its own serde anyway, so `MeteredVersionedKeyValueStoreInternal` would get zero functionality from `MeteredTimestampKeyValueStore`. 

Also, IMO `MeteredVersionedKeyValueStoreInternal` and `MeteredTimestampKeyValueStore` are different conceptually even though they both have the signature `MeteredKeyValueStore<K, ValueAndTimestamp<V>>`. The reason is because the metered layer (in addition to collecting metrics) is responsible for handling serialization to/from the inner bytes stores (pending your other comment, which we can discuss there), and the serialization behavior of the two stores is different, since they use different serdes. This is a more minor reason, though. If you think it's conceptually easier to think of versioned stores as a special case of timestamped stores, we can have `MeteredVersionedKeyValueStoreInternal extends MeteredTimestampKeyValueStore` instead.